### ISSUE TITLE
Mailgun encode_body was written specially for hackney and does not work with finch and others

### ISSUE THREAD

**Issue by abartier** [swoosh/swoosh#763]
### Describe the bug

When sending an email with an attachment using Finch I've got this error in production: 
```
** (ArgumentError) errors were found at the given arguments:
  * 1st argument: not an iodata term
    :erlang.iolist_size({:multipart, [{"from", "\"name\" <hi@example.com>"}, {"subject", "subject"}, {"text", "\n==============================\n  content \n\n==============================\n"}, {"to", "\"name\" <mail@example.com>"}, {"h:Reply-To", "mail@example.com"}, {:file, "/tmp/h60sbk5piax8/b8mwa4pl21fk.pdf", {"form-data", [{"name", "\"attachment\""}, {"filename", "\"name_of_the_file.pdf\""}]}, []}]})
    (mint 1.4.2) lib/mint/http1.ex:980: anonymous fn/1 in Mint.HTTP1.add_content_length_or_transfer_encoding/2
    (mint 1.4.2) lib/mint/core/util.ex:133: Mint.Core.Util.put_new_header_lazy/3
    (mint 1.4.2) lib/mint/http1.ex:981: Mint.HTTP1.add_content_length_or_transfer_encoding/2
    (mint 1.4.2) lib/mint/http1.ex:242: Mint.HTTP1.request/5
    (finch 0.14.0) lib/finch/http1/conn.ex:108: Finch.Conn.request/6
    (finch 0.14.0) lib/finch/http1/pool.ex:45: anonymous fn/8 in Finch.HTTP1.Pool.request/5
    (nimble_pool 0.2.6) lib/nimble_pool.ex:349: NimblePool.checkout!/4
```
When switching to Hackney as api_client everything works as expected 
We need to use the Mailgun adapter on dev to reproduce the bug.

### Steps to Reproduce the Bug or Issue

In a newly created Phoenix app using the default Finch :api_client add a mailer: 

    defmodule MyApp.NameSpace.AttachmentMailer do
      import Swoosh.Email
  
      alias MyApp.Mailer
    
      defp deliver_with_attachment(from, recipient, subject, body, pdf) do
        
        email =
          new()
          |> to({recipient.first_name, recipient.email})
          |> from({from.name, from.email})
          |> reply_to({from.name, from.email})
          |> subject(subject)
          |> text_body(body)
          |> attachment(
            Swoosh.Attachment.new(
              pdf,
              filename: "name_of_the_file.pdf"
            )
          )
    
         with {:ok, _metadata} <- Mailer.deliver(email) do
           {:ok, email}
         end
      end
    end

config/prod.ex

```
 # Configures Swoosh API Client
 config :swoosh, api_client: Swoosh.ApiClient.Finch, finch_name: MyApp.Finch
```
config/runtime.ex
```
  config :my_app, MayApp.Mailer,
    adapter: Swoosh.Adapters.Mailgun,
    api_key: System.get_env("MAILGUN_API_KEY"),
    domain: System.get_env("MAILGUN_DOMAIN")
  #
  # For this example you need include a HTTP client required by Swoosh API client.
  # Swoosh supports Hackney and Finch out of the box:
  #
      config :swoosh, :api_client, Swoosh.ApiClient.Finch

```

### Platform

``` 
elixir -v
 Erlang/OTP 24 [erts-12.3.1] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [jit]
 Elixir 1.14.0 (compiled with Erlang/OTP 23)
```
Phoenix version 1.7.2
Swoosh version 1.9.1

**Comment by princemaple** [comment]
Thanks for reporting.

**Comment by princemaple** [comment]
Laying out what needs to be done to fix this:

- Bring in an implementation of multipart encode, as a helper module. Tesla has a [good one](https://github.com/elixir-tesla/tesla/blob/master/lib/tesla/multipart.ex) that is simple enough to understand and maintain
- Replace Mailgun code to use this multipart helper module, and decouple from the special implementation that only worked with Hackney
- Make sure all still works and replace anything else that may also be depending on hackney details

**Comment by princemaple** [comment]
Found a ready to use solution for multipart encoding: https://github.com/breakroom/multipart

**Comment by krainboltgreene** [comment]
Man it's crazy googling this exact error with "swoosh" didn't bring me to this issue. Damn.

Is there anyway I can help? Also, is there a hack that I can use until we get something merged in?

**Comment by princemaple** [comment]
Hi @krainboltgreene , all the information needed is in my above comment. I just never got the time to work on this. If someone wants to take on this, it'd be great.

It is somewhat too complicated and not worth the effort. I might consider a total rewrite easier using [this mime endpoint instead](https://documentation.mailgun.com/docs/mailgun/api-reference/openapi-final/tag/Messages/#tag/Messages/operation/httpapi.(*apiHandler).handler-fm-19)

There is no easy workaround. Either use hackney or don't use mailgun. Not great, I know :(

**Comment by krainboltgreene** [comment]
I have the time and need to do this, unfortunately. I assume the goal here is to rewrite https://github.com/swoosh/swoosh/blob/main/lib/swoosh/adapters/mailgun.ex#L241-L248 to use https://github.com/breakroom/multipart

**Comment by krainboltgreene** [comment]
Actually I guess this would happen inside the finch API Client.

**Comment by princemaple** [comment]
> I have the time and need to do this, unfortunately. I assume the goal here is to rewrite https://github.com/swoosh/swoosh/blob/main/lib/swoosh/adapters/mailgun.ex#L241-L248 to use https://github.com/breakroom/multipart

This is correct, but also https://github.com/swoosh/swoosh/blob/main/lib/swoosh/adapters/mailgun.ex#L184-L193. This is why I think a rewrite to use the mime api might be better.

> Actually I guess this would happen inside the finch API Client.

No, the client shouldn't care about what it is sending.

**Comment by krainboltgreene** [comment]
```
Because "the lock" specifies mime 2.0.5 and multipart < 0.3.0 depends on mime ~> 1.2, the lock is incompatible with multipart < 0.3.0.
And because your app depends on the lock, multipart < 0.3.0 is forbidden.
So, because your app depends on multipart ~> 0.1.0, version solving failed.
```

Looks like I'm opening two PRs.

**Comment by princemaple** [comment]
Or just copy paste the one from Tesla.

**Comment by krainboltgreene** [comment]
Turns out to have been old documentation, latest version is fine.

**Comment by princemaple** [comment]
Another option is do what `Mua` does. Use `mail` to build the mime. https://github.com/swoosh/swoosh/blob/main/lib/swoosh/adapters/mua.ex#L125

**Comment by princemaple** [comment]
Fixed in #877
